src/state.rs

```diff
@@ -627,17 +635,17 @@ impl DemandProxyState {
     // when we get it, or nothing if the timeout is exceeded, whichever happens first
     pub async fn wait_for_workload(
         &self,
-        addr: &NetworkAddress,
+        wl: &WorkloadInfo,
```

Comment:
The more I think about this, the more I think we can't allow XDS cache lookups by WorkloadInfo directly.

XDS `Workloads` are XDS, and ZDS `WorkloadInfos` are ZDS. They aren't directly convertable, and for inpod you need _both_ pieces to make an authoritative decision about a workload's identity.

Could we maybe use/repurpose `identity/manager.rs`'s `Identity` struct here as the "glue struct"?

- You need a Workload _and_ a WorkloadInfo to _fully_ construct an Identity struct.
- XDS State lookups by Identity struct only (not WLInfo)

Reply:
If you need a Workload to get an Identity, and an Identity to get XDS state, and Workload is XDS state, this is circular -- where do we break the loop (or where do I misunderstand)?

I do think `LocalWorkloadInformation.get_workload` does need to always do a lookup (since it can change at anytime) for sure, instead of caching once at Proxy startup.

Reply:
Sorry - I should have clarified/still thinking.

`no cache lookups by WorkloadInfo directly for local workloads in the _routing path_`

So this is something like

1. on WorkloadInfo (zds), subscribe/get matching XDS entry (matching == same Name/NS/SA)
2. Construct Identity from union of ZDS + XDS (NS/SA/TD to start with, since that's more or less the "important" data union here between these types)
3. Use this to start cert prefetch, assert that any cert we get back has the correct NS/SA/TD we got from the identity/union.
4. From this point on, we can establish mTLS connections safely, and WLInfo basically drops out of use in favor of Identity.
5. Routing path lookups for Workloads keyed are by Identity only

something like that.

Reply:
(this is out of scope of this PR, I am mostly just thinking out loud)

Reply:
> Routing path lookups for Workloads are keyed by Identity/WL only (not WLInfo)


Do we need any lookup at all? Isn't it just `proxy.get_local_workload()`?

Reply:
Probably not, once we have established that `local_workload == local_workloadinfo` - when I say "lookup" I am mostly talking about certs.

Reply:
Part of the goal with this is that you cannot do `lookup(identity)` -- then you can get the wrong one!.

A `Proxy` is scoped to one specific workload. So we already know the identity of it. I want to expose `proxy.get_local_cert()` and `proxy.get_local_workload()` exclusively, so we never do any IP -> cert or IP -> Workload lookups (to identify what we are proxying for; of course for outbound we need a check like that)

Reply:
:+1: yep fair enough - 100% of this is me just not wanting to propagate multiple structs that can be/might be used for identity/RBAC lookup purposes in the hotpath.

I'm imagining something where WorkloadInfo might have information used to establish/validate the `local_workload`'s identity or validity - we already sort of do this in RBAC checks where we compare WLInfo and Workload and effectively use the union of both to make a decision. I'd rather keep away from stuff like that and just convert to something authoritative as early in the local workload's lifecycle as possible.

(but this is orthogonal to what you're doing here so out-of-scope)